PDF/A-3b support

jedbrown opened the pull request:
The resulting papers verify as PDF/A-3b with `verapdf`, but there are some issues left before we can merge:

- [x] I needed to switch to lualatex for this, which would be fine except it isn't working with `babel` (works in simple tests, however). Perhaps someone (@tarleb?) knows how to fix this. `pdflang` will be set automatically once `babel` works.
- [x] For some reason, `author-meta` is not being provided by Pandoc, thus `hyperxmp` is falling bock to the `\author` block, which mangles the author list.
- [x] There are a number of additional fields that can be set. See the [hyperxmp docs](https://mirror.las.iastate.edu/tex-archive/macros/latex/contrib/hyperxmp/hyperxmp.pdf).
- [ ] Inclusion of some images will result in lack of compliance. I haven't figured out why. Some included PDF (non-PDF/A) work but some don't. It appears that PDF created by `rsvg-convert` (i.e., cairo) prevent verification, though PDF created by TeX or ImageMagick (ugly rasters) are fine. Unfortunately, the issue with Cairo also means that `![](image.svg)` results in a PDF/A that fails to verify. This might need to be fixed upstream in librsvg/cairo.

<details>

```xml
      <validationReport profileName="PDF/A-3B validation profile" statement="PDF file is not compliant with Validation Profile requirements." isCompliant="false">
        <details passedRules="123" failedRules="1" passedChecks="112968" failedChecks="4">
          <rule specification="ISO 19005-3:2012" clause="6.2.8" testNumber="3" status="failed" passedChecks="0" failedChecks="4">
            <description>If an Image dictionary contains the Interpolate key, its value shall be false.
            For an inline image, the I key shall have a value of false.</description>
            <object>PDXImage</object>
            <test>Interpolate == false</test>
            <check status="failed">
              <context>root/document[0]/pages[1](114 0 obj PDPage)/contentStream[0](115 0 obj PDContentStream)/operators[507]/xObject[0]/contentStream[0](110 0 obj PDContentStream)/operators[
12]/xObject[0]/contentStream[0](122 0 obj PDContentStream)/operators[3]/xObject[0]/contentStream[0](135 0 obj PDContentStream)/operators[2]/xObject[0]/contentStream[0](143 0 obj PDContentStre
am)/operators[9]/xObject[0](163 0 obj PDXImage)</context>
            </check>
            <check status="failed">
              <context>root/document[0]/pages[1](114 0 obj PDPage)/contentStream[0](115 0 obj PDContentStream)/operators[507]/xObject[0]/contentStream[0](110 0 obj PDContentStream)/operators[
17]/xObject[0]/contentStream[0](123 0 obj PDContentStream)/operators[3]/xObject[0]/contentStream[0](137 0 obj PDContentStream)/operators[2]/xObject[0]/contentStream[0](145 0 obj PDContentStre
am)/operators[9]/xObject[0](165 0 obj PDXImage)</context>
            </check>
            <check status="failed">
              <context>root/document[0]/pages[1](114 0 obj PDPage)/contentStream[0](115 0 obj PDContentStream)/operators[507]/xObject[0]/contentStream[0](110 0 obj PDContentStream)/operators[
22]/xObject[0]/contentStream[0](124 0 obj PDContentStream)/operators[3]/xObject[0]/contentStream[0](139 0 obj PDContentStream)/operators[2]/xObject[0]/contentStream[0](147 0 obj PDContentStre
am)/operators[9]/xObject[0](167 0 obj PDXImage)</context>
            </check>
            <check status="failed">
              <context>root/document[0]/pages[1](114 0 obj PDPage)/contentStream[0](115 0 obj PDContentStream)/operators[507]/xObject[0]/contentStream[0](110 0 obj PDContentStream)/operators[
27]/xObject[0]/contentStream[0](125 0 obj PDContentStream)/operators[3]/xObject[0]/contentStream[0](141 0 obj PDContentStream)/operators[2]/xObject[0]/contentStream[0](149 0 obj PDContentStre
am)/operators[9]/xObject[0](169 0 obj PDXImage)</context>
            </check>
          </rule>
        </details>
      </validationReport>
```
</details>

Cc: https://github.com/openjournals/joss/issues/901 https://github.com/openjournals/joss/issues/1030

tarleb commented:
Thank you, this is great! I'll have a look at the first two points on Monday.

tarleb commented:
I've updated the published pandoc Docker images and made some minor changes to the template, which allowed me to compile this in the image. It's late here, so I'm just dumping the diff here (sorry)

``` diff
diff --git a/Dockerfile b/Dockerfile
index bd3c1f8..fab0d78 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,4 +1,4 @@
-FROM pandoc/latex:2.16.2
+FROM pandoc/latex:edge-alpine
 
 RUN apk add --no-cache ttf-hack
 
@@ -8,7 +8,6 @@ RUN tlmgr update --self && tlmgr install \
   algorithms \
   booktabs \
   caption \
-  collection-xetex \
   environ \
   etoolbox \
   fancyvrb \
diff --git a/data/templates/default.latex b/data/templates/default.latex
index d070f84..4f12290 100644
--- a/data/templates/default.latex
+++ b/data/templates/default.latex
@@ -10,14 +10,29 @@
 \usepackage{hyperxmp}
 \usepackage{embedfile}
 \hypersetup{%
+    unicode=true,
     pdfapart=3,
     pdfaconformance=B,
-    pdfpublication={Journal of Open Source Software},
+    $if(title-meta)$
+    pdftitle={$title-meta$},
+    $endif$
+    $if(author-meta)$
+    pdfauthor={$author-meta$},
+    $endif$
+    $if(keywords)$
+    pdfkeywords={$for(keywords)$$keywords$$sep$; $endfor$},
+    $endif$
+    pdfpublication={$journal_name$},
     pdfpublisher={Open Journals},
-    colorlinks,
-    breaklinks=true,
-    urlcolor=[rgb]{0.0, 0.5, 1.0},
-    linkcolor=[rgb]{0.0, 0.5, 1.0}
+    $if(colorlinks)$
+    colorlinks=true,
+    linkcolor=$if(linkcolor)$$linkcolor$$else$Maroon$endif$,
+    citecolor=$if(citecolor)$$citecolor$$else$Blue$endif$,
+    urlcolor=$if(urlcolor)$$urlcolor$$else$Blue$endif$,
+    $else$
+    pdfborder={0 0 0},
+    $endif$
+    breaklinks=true
 }
 % https://tex.stackexchange.com/a/535849
 \embedfile[afrelationship={/Source},ucfilespec={\jobname.tex},mimetype={application/x-tex}]{\jobname.tex}
@@ -265,27 +280,8 @@ $endif$
 $if(colorlinks)$
 \PassOptionsToPackage{usenames,dvipsnames}{color} % color is loaded by hyperref
 $endif$
-\hypersetup{unicode=true,
-$if(title-meta)$
-            pdftitle={$title-meta$},
-$endif$
-$if(author-meta)$
-            pdfauthor={$author-meta$},
-$endif$
-$if(keywords)$
-            pdfkeywords={$for(keywords)$$keywords$$sep$; $endfor$},
-$endif$
-$if(colorlinks)$
-            colorlinks=true,
-            linkcolor=$if(linkcolor)$$linkcolor$$else$Maroon$endif$,
-            citecolor=$if(citecolor)$$citecolor$$else$Blue$endif$,
-            urlcolor=$if(urlcolor)$$urlcolor$$else$Blue$endif$,
-$else$
-            pdfborder={0 0 0},
-$endif$
-            breaklinks=true}
 \urlstyle{same}  % don't use monospace font for urls
-$if(lang-babel-breaks-with-luatex)$
+$if(lang)$
 \ifLuaTeX
 \usepackage[bidi=basic]{babel}
 \else
```

tarleb commented:
I forgot to mention: please wait 20 min before trying this, as to allow the new pandoc image to be built (see [here](https://github.com/pandoc/dockerfiles/actions/runs/1679313889))

tarleb commented:
The `author-meta` field is now also set, but it shows up only in the normal pdf metadata, but not in pdfa.

jedbrown commented:
Thanks for working on this. I applied the patch, but I'm still having trouble
```console
$ make docker-image
$ docker run -it --rm -v /path/to/papers/joss:/data -u $(id -u):$(id -g) openjournals/inara -o pdf paper.md
[...]
Latexmk: Run number 1 of rule 'lualatex'
Collected error summary (may duplicate other messages):
  lualatex: Command for 'lualatex' gave return code 1
      Refer to '/tmp/tex2pdf.-312a5980667ef9e3/input.log' for details
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs, or warnings treated as errors.
[makePDF] Run #1
This is LuaHBTeX, Version 1.13.2 (TeX Live 2021)
 restricted system commands enabled.


luaotfload | load : FATAL ERROR
luaotfload | load :   × Failed to load "fontloader" module "basics-gen".
luaotfload | load :   × Error message:
luaotfload | load :     × "...dir/texmf-dist/tex/luatex/luaotfload/luaotfload-init.lua:301: system : no writeable cache path, quiting".

stack traceback:
        ...e/texdir/texmf-dist/tex/luatex/luaotfload/luaotfload.lua:206: in local 'load_fontloader_module'
        ...dir/texmf-dist/tex/luatex/luaotfload/luaotfload-init.lua:303: in upvalue 'init_main'
        ...dir/texmf-dist/tex/luatex/luaotfload/luaotfload-init.lua:584: in function 'luaotfload-init.lua'
        ...e/texdir/texmf-dist/tex/luatex/luaotfload/luaotfload.lua:330: in field 'main'
        [\directlua]:1: in main chunk


Error producing PDF.
!  ==> Fatal error occurred, no output PDF file produced!
```

tarleb commented:
Not super pretty, but adding this to the end of the `Dockerfile` works:

``` Dockerfile
RUN TERM=dumb luaotfload-tool --update \
  && chmod -R o+w /opt/texlive/texdir/texmf-var
```

jedbrown commented:
Are you sure `author-meta` is set? I get an error if I remove the guards and the output is mangled because I'm pretty sure `hyperxmp` recognizes that it's not set explicitly and falls back to extracting it from `\author`, which has affiliation stuff in it. For example:  `exempi -x paper.pdf`

```xml
   <dc:creator>
    <rdf:Seq>
     <rdf:li>Jed Brown1</rdf:li>
     <rdf:li>Ahmad Abdelfattah3</rdf:li>
     <rdf:li>Valeria Barra1</rdf:li>
     <rdf:li>Natalie Beams3</rdf:li>
     <rdf:li>Jean-Sylvain Camier2</rdf:li>
     <rdf:li>Veselin Dobrev2</rdf:li>
     <rdf:li>Yohann Dudouit2</rdf:li>
     <rdf:li>Leila Ghaffari1</rdf:li>
     <rdf:li>Tzanio Kolev2</rdf:li>
     <rdf:li>David Medina4</rdf:li>
     <rdf:li>Will Pazner2</rdf:li>
     <rdf:li>Thilina Ratnayaka5</rdf:li>
     <rdf:li>Jeremy Thompson1</rdf:li>
     <rdf:li>and Stan Tomov3[]</rdf:li>
     <rdf:li>ulm</rdf:li>
     <rdf:li>elm</rdf:li>
     <rdf:li>lm</rdf:li>
     <rdf:li>slm</rdf:li>
     <rdf:li>mm</rdf:li>
     <rdf:li>sbm</rdf:li>
     <rdf:li>bm</rdf:li>
     <rdf:li>ebm</rdf:li>
     <rdf:li>ubm</rdf:li>
     <rdf:li>muc</rdf:li>
     <rdf:li>mec</rdf:li>
     <rdf:li>mc</rdf:li>
     <rdf:li>msc</rdf:li>
     <rdf:li>msx</rdf:li>
     <rdf:li>mx</rdf:li>
     <rdf:li>mex</rdf:li>
     <rdf:li>mux</rdf:li>
    </rdf:Seq>
   </dc:creator>
```

tarleb commented:
The fix for that is on the `main` branch; I'll push the fix above, too. Can you rebase and try again in a couple of minutes?

jedbrown commented:
Rebased and added the extra fields. I think the issue with PDF images can be addressed later, perhaps by running `verapdf` and alerting the author. I use librsvg a lot, but I think it's not that common among our papers and merging this will get us PDF/A for most papers.

jedbrown commented on data/templates/default.latex:
```diff
@@ -467,7 +467,7 @@ $endif$
     \setlength\itemsep{0em}
     \item \href{$software_review_url$}{\color{linky}{Review}} \ExternalLink
     \item \href{$software_repository_url$}{\color{linky}{Repository}} \ExternalLink
-    \item \href{https://dx.doi.org/$archive_doi$}{\color{linky}{Archive}} \ExternalLink
+    \item \href{https://doi.org/$archive_doi$}{\color{linky}{Archive}} \ExternalLink
```
This commit should be merged soon even if the rest of the commit needs more review.

jedbrown commented on data/templates/default.latex:
```diff
@@ -1,15 +1,66 @@
+    $if(keywords)$
+    pdfkeywords={$for(keywords)$$keywords$$sep$; $endfor$},
+    $endif$
+    pdfpublication={$journal_name$},
+    pdfpublisher={Open Journals},
+    pdfissn={$journal_issn$},
+    pdfpubtype={journal},
+    pdfvolumenum={$volume$},
+    pdfissuenum={$issue$},
+    $if(formatted_doi)$
+    pdfdoi={$formatted_doi$},
+    $endif$
+    $if(author-meta)$
+    pdfcopyright={Copyright (c) $year$, $author-meta$},
+    $endif$
```
I don't know which of these variables need to be treated as optional and which can just be used.

tarleb commented:
Awesome

As you are now the resident PDF/A expert, is it ok if I ping you if we run into problems in the future?

jedbrown commented:
Lol, I'll try. Is this (`main`) now in production? At what point can we announce this?

tarleb commented:
@arfon can probably give a more accurate answer, but I believe the whedon to buffy transition has to happen before this goes into production.

arfon commented:
> @arfon can probably give a more accurate answer, but I believe the whedon to buffy transition has to happen before this goes into production.

Correct. Very soon now I hope (weeks not months)
